### ISSUE TITLE
Example for ECS (Fargate) Run Task without a Service

### ISSUE THREAD

**Issue by burtoni** [aws-samples/aws-cdk-examples#289]
It's not immediately clear from the examples how to run a Fargate Task with multiple concurrent containers without a service. I'm looking for a replacement for the [`aws ecs run-task`](https://docs.aws.amazon.com/cli/latest/reference/ecs/run-task.html) CLI command, allowing for a one-off task to be initiated and executed and including a `--taskCount` parameter.

Is this possible using the CDK in TypeScript?

**Comment by richardhboyd** [comment]
I don’t think CDK supports that. CDK is meant for defining the tasks and setting up the infrastructure to support them, executing tasks is generally a runtime behavior.

I suppose an awscli custom resource could be used if the task is needed as part of a provisioning process, but I would be suspect of such an implementation.

**Comment by burtoni** [comment]
Thanks for the quick response @richardhboyd .

That makes sense, although to be pedantic the Task is infrastructure, just short-lived infrastructure!

I was able to create a Task behind a Service (though obviously the Service will keep restarting the Task once it completes, assuming I want it to be always available), so I assumed the other use case would be supported too.

**Comment by richardhboyd** [comment]
If we're going to be pedantic, the Task is definitely infrastructure, executing it is not. CDK will allow you to create a Task, but it'll take some extra work to execute it as part of your infrastructure provisioning process.

`aws ecs run-task` takes a `--task-definition` property which points to the ARN for a provisioned Task.

TaskDefinitions can be created with the [TaskDefinition](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ecs.TaskDefinition.html) Construct. How you choose to run those tasks is up to you, a Servcie is one option but there's nothign stopping you from using the AWSCustomResource Construct to run the `run-task` operation.

**Comment by burtoni** [comment]
Let me explore the custom resource construct but it may be sufficient to create the task definition in CDK (which has some other nice features for my use case) and use the aws cli to run it.

Apologies for the throwaway pedantic comment, was a poorly chosen phrase.

**Comment by ctrlplusb** [comment]
@burtoni I am needing the same thing. Would be interested to hear your final solution to this problem.

**Comment by marioizquierdo** [comment]
Running the Task once as part of the CDK deploy process is similar to allowing set the `--taskCount` from CDK. If you can run/start the Service from CDK, why not run/start a one-off task as well?

This is needed in some scenarios to allow a full deploy without human intervention. For example, if you have an RDS DB cluster, you may need to create a user with limited privileges to connect from the service. If you deploy the service, it will get stuck trying to start the service tasks because they can't connect with that user. To unblock the stack deploy, you have to manually connect to the DB and use SQL to create the user. If instead, the user could be created from a one-off task, this would not be necessary.

**Comment by Mahoney** [comment]
To add to the DB migration use case - I've ended up adding a sidecar container to the app task that runs before the main container, with the main container requiring it to succeed, but that seems pretty wasteful - we now redundantly run the migration over and over again whenever a task is spun up, slowing task startup.

I experimented with an `AWSCustomResource` using  https://github.com/pahud/cdk-fargate-run-task and it works OK, but it has one major negative which is that it can only kick the task off - it can't block until it completes, nor can it fail the `cdk deploy` if the task exits with a non-zero status. Whereas the sidecar container explicitly supports both using the ContainerDefinition DependsOn Condition.

I'm sure I could write some code to do all of that in a lambda but it feels like a sufficiently generic use case that something in the CDK / CloudFormation could support it. Perhaps even at the DB level - could a `rds.DatabaseInstance` be associated with an ECR image responsible for initialising / migrating it?
